File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    private static boolean[] visited;
    private static List<String> result = new ArrayList<>();
    
    
    public String[] solution(String[][] tickets) {
        
        visited = new boolean[tickets.length];
        
        dfs(0, "ICN", "ICN", tickets);
        Collections.sort(result);
        
        String[] answer = result.get(0).split(" ");
        return answer;
    }
    public static void dfs(int idx, String start, String route, String[][] tickets){
        if(idx == tickets.length){
            result.add(route);
            
            return;
                
        }
        
        for(int i = 0; i < tickets.length; i++){
            if(tickets[i][0].equals(start) && !visited[i]) {
                visited[i] = true;
                dfs(idx + 1, tickets[i][1], route + " " + tickets[i][1], tickets);
                visited[i] = false;
                
            }
        }
        return;
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import defaultdict
def solution(tickets):
    td = defaultdict(list)
    
    for s,e in tickets:
        td[s].append(e)
    
    for tk in td.keys():
        td[tk].sort(reverse = True)
        
    answer = []
    path = ["ICN"]
    
    while path:
        now = path[-1]
        
        if len(td[now]) ==0:
            answer.append(path.pop())
        else:
            path.append(td[now].pop())
    return answer[::-1]